File: ios/TechBlogCatchup/Services/APIClient.swift
import Foundation

struct CrawlResponse: Codable {
    let status: String
    let jobId: Int
    let source: String?

    enum CodingKeys: String, CodingKey {
        case status, source
        case jobId = "job_id"
    }
}

struct GenerateResponse: Codable {
    let status: String
    let jobId: Int
    let postId: Int?
    let limit: Int?

    enum CodingKeys: String, CodingKey {
        case status, limit
        case jobId = "job_id"
        case postId = "post_id"
    }
}

private struct ServerErrorBody: Codable {
    let detail: String?
}

@Observable
final class APIClient {
    static let shared = APIClient()

    private let session: URLSession
    private let decoder: JSONDecoder

    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = AppConfig.requestTimeout
        self.session = URLSession(configuration: config)

        self.decoder = JSONDecoder()
        self.decoder.dateDecodingStrategy = .custom { decoder in
            let container = try decoder.singleValueContainer()
            let dateString = try container.decode(String.self)

            // Try ISO8601 with fractional seconds (microseconds from Python)
            let iso8601Fractional = ISO8601DateFormatter()
            iso8601Fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            if let date = iso8601Fractional.date(from: dateString) {
                return date
            }

            // Try ISO8601 without fractional seconds
            let iso8601 = ISO8601DateFormatter()
            iso8601.formatOptions = [.withInternetDateTime]
            if let date = iso8601.date(from: dateString) {
                return date
            }

            // Try "yyyy-MM-dd'T'HH:mm:ss" (no timezone)
            let formatter = DateFormatter()
            formatter.locale = Locale(identifier: "en_US_POSIX")
            formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss"
            if let date = formatter.date(from: dateString) {
                return date
            }

            // Try "yyyy-MM-dd'T'HH:mm:ss.SSSSSS" (microseconds, no timezone)
            formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSSSS"
            if let date = formatter.date(from: dateString) {
                return date
            }

            throw DecodingError.dataCorruptedError(
                in: container,
                debugDescription: "Cannot decode date: \(dateString)"
            )
        }
    }

    // MARK: - Core Request

    private func request<T: Decodable>(
        _ endpoint: String,
        method: String = "GET",
        body: (any Encodable)? = nil,
        queryItems: [URLQueryItem]? = nil
    ) async throws -> T {
        var attempt = 0
        var lastError: Error = APIError.unknown

        while attempt < AppConfig.maxRetries {
            do {
                return try await performRequest(endpoint, method: method, body: body, queryItems: queryItems)
            } catch let error as APIError {
                switch error {
                case .notFound, .conflict:
                    throw error
                default:
                    lastError = error
                }
            } catch {
                lastError = error
            }

            attempt += 1
            if attempt < AppConfig.maxRetries {
                let backoffMs = AppConfig.initialBackoffMs * (1 << (attempt - 1))
                try await Task.sleep(nanoseconds: UInt64(backoffMs) * 1_000_000)
            }
        }

        throw lastError
    }

    private func performRequest<T: Decodable>(
        _ endpoint: String,
        method: String,
        body: (any Encodable)? = nil,
        queryItems: [URLQueryItem]?
    ) async throws -> T {
        guard var components = URLComponents(string: AppConfig.baseURL + endpoint) else {
            throw APIError.unknown
        }

        if let queryItems, !queryItems.isEmpty {
            components.queryItems = queryItems
        }

        guard let url = components.url else {
            throw APIError.unknown
        }

        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = method
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if let body {
            let encoder = JSONEncoder()
            urlRequest.httpBody = try encoder.encode(AnyEncodable(body))
        }

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: urlRequest)
        } catch {
            throw APIError.networkError(error)
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.unknown
        }

        switch httpResponse.statusCode {
        case 200..<300:
            do {
                return try decoder.decode(T.self, from: data)
            } catch {
                throw APIError.decodingError(error)
            }
        case 404:
            throw APIError.notFound
        case 409:
            let body = try? decoder.decode(ServerErrorBody.self, from: data)
            throw APIError.conflict(body?.detail ?? "Conflict")
        default:
            let body = try? decoder.decode(ServerErrorBody.self, from: data)
            throw APIError.serverError(
                statusCode: httpResponse.statusCode,
                message: body?.detail ?? "Unknown server error"
            )
        }
    }

    // MARK: - Posts

    func fetchPosts(
        source: String? = nil,
        tag: String? = nil,
        search: String? = nil,
        audioStatus: String? = nil,
        qualityMin: Int? = nil,
        offset: Int = 0,
        limit: Int = 20,
        sort: String = "-published_at"
    ) async throws -> PaginatedPosts {
        var queryItems: [URLQueryItem] = [
            URLQueryItem(name: "offset", value: "\(offset)"),
            URLQueryItem(name: "limit", value: "\(limit)"),
            URLQueryItem(name: "sort", value: sort),
        ]
        if let source { queryItems.append(URLQueryItem(name: "source", value: source)) }
        if let tag { queryItems.append(URLQueryItem(name: "tag", value: tag)) }
        if let search { queryItems.append(URLQueryItem(name: "search", value: search)) }
        if let audioStatus { queryItems.append(URLQueryItem(name: "audio_status", value: audioStatus)) }
        if let qualityMin { queryItems.append(URLQueryItem(name: "quality_min", value: "\(qualityMin)")) }

        return try await request("/api/posts", queryItems: queryItems)
    }

    func fetchPost(id: Int) async throws -> PostDetail {
        return try await request("/api/posts/\(id)")
    }

    func fetchPlaylist(
        source: String? = nil,
        tag: String? = nil,
        search: String? = nil,
        offset: Int = 0,
        limit: Int = 50,
        sort: String = "-published_at"
    ) async throws -> PaginatedPosts {
        var queryItems: [URLQueryItem] = [
            URLQueryItem(name: "offset", value: "\(offset)"),
            URLQueryItem(name: "limit", value: "\(limit)"),
            URLQueryItem(name: "sort", value: sort),
        ]
        if let source { queryItems.append(URLQueryItem(name: "source", value: source)) }
        if let tag { queryItems.append(URLQueryItem(name: "tag", value: tag)) }
        if let search { queryItems.append(URLQueryItem(name: "search", value: search)) }

        return try await request("/api/playlist", queryItems: queryItems)
    }

    // MARK: - Tags & Sources

    func fetchTags() async throws -> [TagInfo] {
        return try await request("/api/tags")
    }

    func fetchSources() async throws -> [SourceInfo] {
        return try await request("/api/sources")
    }

    // MARK: - Status & Crawl

    func fetchStatus() async throws -> StatusInfo {
        return try await request("/api/status")
    }

    func fetchCrawlStatus() async throws -> [CrawlStatusItem] {
        return try await request("/api/crawl-status")
    }

    // MARK: - Jobs

    func fetchJobs(
        jobType: String? = nil,
        status: String? = nil,
        limit: Int = 50
    ) async throws -> [JobInfo] {
        var queryItems: [URLQueryItem] = [
            URLQueryItem(name: "limit", value: "\(limit)"),
        ]
        if let jobType { queryItems.append(URLQueryItem(name: "job_type", value: jobType)) }
        if let status { queryItems.append(URLQueryItem(name: "status", value: status)) }

        return try await request("/api/jobs", queryItems: queryItems)
    }

    func fetchJob(id: Int) async throws -> JobInfo {
        return try await request("/api/jobs/\(id)")
    }

    // MARK: - Actions

    func triggerCrawl(source: String? = nil) async throws -> CrawlResponse {
        struct CrawlBody: Encodable {
            let source: String?
        }
        return try await request("/api/crawl", method: "POST", body: CrawlBody(source: source))
    }

    func triggerGenerate(postId: Int? = nil, limit: Int = 10) async throws -> GenerateResponse {
        struct GenerateBody: Encodable {
            let postId: Int?
            let limit: Int

            enum CodingKeys: String, CodingKey {
                case postId = "post_id"
                case limit
            }
        }
        return try await request("/api/generate", method: "POST", body: GenerateBody(postId: postId, limit: limit))
    }
}

// MARK: - Type Erasure Helper

private struct AnyEncodable: Encodable {
    private let _encode: (Encoder) throws -> Void

    init(_ wrapped: any Encodable) {
        self._encode = wrapped.encode
    }

    func encode(to encoder: Encoder) throws {
        try _encode(encoder)
    }
}

File: android/lib/services/api_client.dart
import 'package:dio/dio.dart';
import '../models/post.dart';
import '../models/tag.dart';
import '../models/source.dart';
import '../models/job.dart';
import '../models/paginated_posts.dart';
import '../models/crawl_status.dart';
import '../models/status_info.dart';
import 'api_error.dart';

class ApiClient {
  final Dio _dio;
  final String baseUrl;

  ApiClient({required this.baseUrl})
      : _dio = Dio(BaseOptions(
          baseUrl: baseUrl,
          connectTimeout: const Duration(seconds: 10),
          receiveTimeout: const Duration(seconds: 10),
          headers: {'Content-Type': 'application/json'},
        )) {
    _dio.interceptors.add(_RetryInterceptor(_dio));
  }

  String getAudioUrl(String audioPath) => '$baseUrl/$audioPath';

  // -- Posts --

  Future<PaginatedPosts> getPosts({
    String? source,
    String? tag,
    String? search,
    String? audioStatus,
    int? qualityMin,
    int offset = 0,
    int limit = 20,
    String? sort,
  }) async {
    final params = <String, dynamic>{
      'offset': offset,
      'limit': limit,
    };
    if (source != null) params['source'] = source;
    if (tag != null) params['tag'] = tag;
    if (search != null && search.isNotEmpty) params['search'] = search;
    if (audioStatus != null) params['audio_status'] = audioStatus;
    if (qualityMin != null) params['quality_min'] = qualityMin;
    if (sort != null) params['sort'] = sort;

    final data = await _get('/api/posts', params: params);
    return PaginatedPosts.fromJson(data as Map<String, dynamic>);
  }

  Future<PostDetail> getPost(int id) async {
    final data = await _get('/api/posts/$id');
    return PostDetail.fromJson(data as Map<String, dynamic>);
  }

  Future<PaginatedPosts> getPlaylist({
    int offset = 0,
    int limit = 20,
    String? sort,
  }) async {
    final params = <String, dynamic>{
      'offset': offset,
      'limit': limit,
    };
    if (sort != null) params['sort'] = sort;

    final data = await _get('/api/playlist', params: params);
    return PaginatedPosts.fromJson(data as Map<String, dynamic>);
  }

  // -- Tags & Sources --

  Future<List<Tag>> getTags() async {
    final data = await _get('/api/tags');
    return (data as List).map((e) => Tag.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<List<Source>> getSources() async {
    final data = await _get('/api/sources');
    return (data as List).map((e) => Source.fromJson(e as Map<String, dynamic>)).toList();
  }

  // -- Status --

  Future<StatusInfo> getStatus() async {
    final data = await _get('/api/status');
    return StatusInfo.fromJson(data as Map<String, dynamic>);
  }

  Future<List<CrawlStatusItem>> getCrawlStatus() async {
    final data = await _get('/api/crawl-status');
    return (data as List)
        .map((e) => CrawlStatusItem.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<Map<String, dynamic>> getHealth() async {
    final data = await _get('/api/health');
    return data as Map<String, dynamic>;
  }

  // -- Jobs --

  Future<List<Job>> getJobs({String? jobType, String? status}) async {
    final params = <String, dynamic>{};
    if (jobType != null) params['job_type'] = jobType;
    if (status != null) params['status'] = status;

    final data = await _get('/api/jobs', params: params);
    return (data as List).map((e) => Job.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<Job> getJob(int id) async {
    final data = await _get('/api/jobs/$id');
    return Job.fromJson(data as Map<String, dynamic>);
  }

  // -- Actions --

  Future<Map<String, dynamic>> triggerCrawl({String? source}) async {
    final body = <String, dynamic>{};
    if (source != null) body['source'] = source;
    final data = await _post('/api/crawl', body: body);
    return data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> triggerGenerate({int? postId, int limit = 10}) async {
    final body = <String, dynamic>{'limit': limit};
    if (postId != null) body['post_id'] = postId;
    final data = await _post('/api/generate', body: body);
    return data as Map<String, dynamic>;
  }

  // -- Internal --

  Future<dynamic> _get(String path, {Map<String, dynamic>? params}) async {
    try {
      final response = await _dio.get(path, queryParameters: params);
      return response.data;
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  Future<dynamic> _post(String path, {Map<String, dynamic>? body}) async {
    try {
      final response = await _dio.post(path, data: body);
      return response.data;
    } on DioException catch (e) {
      throw _handleError(e);
    }
  }

  ApiError _handleError(DioException e) {
    if (e.response != null) {
      final data = e.response!.data;
      String message = '${e.response!.statusCode} ${e.response!.statusMessage}';
      String? details;

      if (data is Map<String, dynamic>) {
        if (data['detail'] is String) {
          details = data['detail'] as String;
          message = details;
        } else if (data['message'] is String) {
          details = data['message'] as String;
          message = details;
        }
      }

      return ApiError(
        statusCode: e.response!.statusCode ?? 0,
        message: message,
        details: details,
      );
    }

    if (e.type == DioExceptionType.connectionTimeout ||
        e.type == DioExceptionType.receiveTimeout) {
      return const ApiError(statusCode: 0, message: 'Request timed out');
    }

    return const ApiError(statusCode: 0, message: 'Network error');
  }
}

class _RetryInterceptor extends Interceptor {
  final Dio _dio;
  static const _maxRetries = 3;
  static const _initialBackoffMs = 500;

  _RetryInterceptor(this._dio);

  @override
  void onError(DioException err, ErrorInterceptorHandler handler) async {
    try {
      final statusCode = err.response?.statusCode ?? 0;
      final isRetryable = statusCode >= 500 || statusCode == 429;

      if (!isRetryable) {
        handler.next(err);
        return;
      }

      final extra = err.requestOptions.extra;
      final attempt = (extra['_retryAttempt'] as int?) ?? 0;

      if (attempt >= _maxRetries - 1) {
        handler.next(err);
        return;
      }

      final delay = _initialBackoffMs * (1 << attempt);
      await Future.delayed(Duration(milliseconds: delay));

      final options = err.requestOptions;
      options.extra['_retryAttempt'] = attempt + 1;

      final response = await _dio.fetch(options);
      handler.resolve(response);
    } on DioException catch (e) {
      handler.next(e);
    } catch (_) {
      handler.next(err);
    }
  }
}
